### PR TITLE
Clean slate, initial implementation with tests

### PR DESCRIPTION
**Pull request by RangerMauve**

### CODE REVIEW

**Comment by gmaclennan** [comment]
@RangerMauve I'm just looking through this, can you ensure that the tests actually test that sync works (e.g. that a document and an image stored on each peer are available on the other peer after sync)? Reading through the code it looks to me like `replicateFromWebsocket()` will connect to a peer, but will not initiate sync. The code in mapeo-core is in dire need of a re-write, but the command that actually kicks off sync is [`peer.handshake.accept()`](https://github.com/digidem/mapeo-core/blob/ws-sync/sync.js#L306). This property (`peer.handshake`) is added by the [`addPeer()`](https://github.com/digidem/mapeo-core/blob/ws-sync/sync.js#L568) method, and I don't think sync starts until the handshake is accepted. This is flakey code anyway, because `peer.handshake` is added and then removed, and if the user tries to sync when it is `undefined`, then sync fails.

**Comment by RangerMauve** [comment]
Sweet, thank you for the input. I'll make sure to take that into account.

Would creating data on one side and trying to load it on the other side after sync suffice?

**Comment by gmaclennan** [comment]
Yes, see the sync tests in mapeo-core for some exampes about how to do this: https://github.com/digidem/mapeo-core/blob/ws-sync/test/sync.js#L214

File: README.md
Comment (gmaclennan) [inline README.md:98]: What's the default timeout on http connection for nginx? Sync could take 30 mins or more over a slow connection, so we want to ensure nginx would not timeout.

File: bin.js
Comment (gmaclennan) [inline bin.js]: Nitpik, isn't this the delay until instances are released for gc, but gc could happen much later?

File: index.js
Comment (gmaclennan) [inline index.js]: I think we should treat this as the idea of creating a "project" record on the server. We may want to add more refined permissions in the future. I think it would also be good practice to not expose the project key on any public end point, e.g. key is write only, projectId === discovery key and is used for public transactions.

Therefore: path could be `/projects/` (plural to follow common REST semantics)
Create a new project on the server, POST to `/projects/` with `key` in body.
Returns project ID (=== discovery key)
Delete a project, DELETE `/projects/:discoveryKey`.

Initially, existence of project record === permission to sync.
In the future: might add more details to a project record to indicate refined permissions (e.g. resource limits).

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js:55]: Let's use the discovery key for this. I want to avoid leaving secret keys exposed where possible.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]: It would be good, especially during our initial piloting of this service, to log all sync progress somewhere, so that we can get an idea of how it is working and what performance issues there are. Do you have experience with logging services? The one I know of is https://apex.sh/logs/ but I've never used it.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js:30]: Yes, let's have a single place, e.g. a leveldb, where secret project keys are written, and use the derived discovery key for the identifier in all other places in the app.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]: I'm worried this logic is a little hard to parse and could result in unexpected behaviour.

Each peer that connects is setting a timeout, and closing an instance will happen if the timeout happens at a moment when there are no peers connected.

I think the expected logic would be: instance is closed gcTimeout seconds after the last peer finished syncing. I don't think that behaviour is guaranteed by this code right now.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]:
```suggestion
    const total = this.instances.size
```
?

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js:169]: For the all the issues with promises, this kind of code is so much easier as a promise:

```js
  close () {
    const instances = Array.from(this.instances.values())
    return Promise.all(instances.map(mapeo => promisify(mapeo.close))
  }
```

But fine to keep with callback syntax for now. We plan to [switch mapeo-core to using promises](https://github.com/digidem/mapeo-rfc/blob/995ad997ddfcd3c996c8a9e6b4c570f6ee88b777/RFC-003-mapeo-core.md#12-callback-based).

File: permissions.js
Comment (gmaclennan) [inline permissions.js]: nitpick: this and below is a `get` not a `has`.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
This is looking really great. I added a few comments, some nitpicks. For future maintenance, I would really like to reduce how much we pass around the secret keys for projects. Currently we use them both as an identifier (and dir name). I think we should think of the server state being a list of "projects" that are created on the server, identified by their discovery key, and each project currently only has one property (`canSync=true`), which currently is implied: a project exists, then it can sync. Then there should be a single place that stores the secret keys, which can be looked up by project ID (discovery key). The code mainly does this, but it seems like the secret keys leak out in a few places.

**Comment by gmaclennan** [comment]
What tasks are remaining to get this to a place where we are ready to pilot it?

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js]: I was thinking it'd be to STDOUT and to use journald for managing them. Is using an external logging service a hard requirement?

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js:169]: Yeah, I usually use promises, but didn't want to stray too far from mapeo-core's code style.

**Comment by RangerMauve** [comment]
@gmaclennan Here's what I think is left to pilot it:

- Update the code per your review
- Update mapeo-core in my branches of mapeo-desktop and mapeo-mobile
- Double check that it's working
- Generate a new mapeo settings file (currently using the workshop one for testing which is not ideal)
- Get builds of the app for people to install

I'm guessing we could have all that by next week.

File: README.md
Comment (RangerMauve) [reply README.md:98]: Good call, what should I set it to?

Right now it looks like the timeout is there for when there are no bytes being sent or received over a 60 second interval.

If the sync is happening for 30 mins or more and is actually sending data during that time, the default config should work.

https://developpaper.com/detailed-explanation-of-nginx-timeout-configuration/

```
server {
   ...
   proxy_read_timeout 300;
   proxy_connect_timeout 300;
   proxy_send_timeout 300;
   ...
}
```

File: index.js
Comment (RangerMauve) [reply index.js]: Is having the ID in the response necessary?

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js:30]: Only caveat is that `multi.get()` needs the project key so it can initialize the mapeo instance properly.

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js]: Are there peers that stay connected after they've finished syncing?

**Comment by RangerMauve** [comment]
Made some progress on the review, got some questions for clarification for when you have time @gmaclennan

File: README.md
Comment (gmaclennan) [reply README.md:98]: ok good. Just to be sure, I think a 5 minute keep-alive is good. I have seen satellite internet with > 60 second ping times. I think by the time you get to 5 minutes without data, you may as well give up trying to sync!

File: index.js
Comment (gmaclennan) [reply index.js]: If we build a front-end "admin" UI for this (to add projects to Mapeo web and configure their permissions), then it is useful get back the ID when you create a new object. The client could assume the ID of a newly created project resource would be the discovery key, which the client could also derive, but if we return the ID then we don't hold ourselves to that (a project ID could be something different in the future)

File: multi-mapeo.js
Comment (gmaclennan) [reply multi-mapeo.js]: No, I think that would be a good solution. We'd have to keep an eye on things, e.g. how long before they could take up loads of disk space? We would need to have our own way of analysing the logs and querying them. I have _zero_ experience in all this!

File: multi-mapeo.js
Comment (gmaclennan) [reply multi-mapeo.js]: I don't think so. I don't think I was clear, and maybe I am mis-understanding the code here, but this is the scenario I am thinking:

GC_TIMEOUT = 600s

0s: Peer A connects, starts to sync
100s: Peer A finishes sync, disconnects, timeout A is started.
...
500s: Peer B connects, starts sync
650s: Peer B finishes sync, disconnects, timeout B is started.
700s: timeout A completes, no connections, instance closed.

Here, the instance is closed 50s after the last peer finished sync. I would expect the instance to remain active as long as peers continue to connect and sync within the GC_TIMEOUT time frame.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]: This could be called twice, because we are not tracking the `closing` state of a mapeo instance. E.g. in line 73, `instances.has(key)` could return true between the `close()` method being called and the callback.
Small window, but a source of weird bugs down the line.

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js]: Ah I understand now, thank you. I think a good approach there would be to clear any existing timeouts before starting a new one.

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js]: Good catch

File: multi-mapeo.js
Comment (RangerMauve) [reply multi-mapeo.js]: Is this something we should have before this PR gets merged, or could we have it as a TODO for after the initial testing?

Comment (RangerMauve) [reply multi-mapeo.js]: I've added some basic logging using fastify and pino, so we should at least see things replicating.

**Comment by RangerMauve** [comment]
Made it a bit more robust / usable.

**Comment by RangerMauve** [comment]
@gmaclennan Could I get a final review on this too please? 😁

File: index.js
Comment (gmaclennan) [inline index.js:46]: How do errors work in Websocket? Is there a way to indicate an error to the client? Should this be JSON so that it can be parsed correctly on the client? How do we handle this error in client code? Do we test for this error (I think it is a likely error to happen)?

File: index.js
Comment (gmaclennan) [inline index.js]: We don't need to return anything from this function for a websocket request right? No need for this to be an async function?

File: index.js
Comment (gmaclennan) [inline index.js]: Should set status code, and error should be JSON and documented.

File: index.js
Comment (gmaclennan) [inline index.js:84]: JSON error.

File: index.js
Comment (gmaclennan) [inline index.js:88]: JSON error. Set status code.

File: index.js
Comment (gmaclennan) [inline index.js]: Should validate body, easy to do with fastify with a schema for req.body.

File: index.js
Comment (gmaclennan) [inline index.js:97]: JSON error and status code.

File: index.js
Comment (gmaclennan) [inline index.js]: This is a dangerous end point to have, it leaks secret keys for everyone. What do we need it for? Could we list discovery keys instead? I feel like project keys should never be exposed once added.

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]: Should not log project keys

File: multi-mapeo.js
Comment (gmaclennan) [inline multi-mapeo.js]: don't log project key

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Hi Mauve, I've left a few comments in the code, but there are a few areas I think need to be adjusted before merging this:

1. Errors: I think we should return error status codes (e.g. 500) and the body of an error response should be JSON (I'm not sure what it is now, but looks like the code is sending an error as the message body? How should the client deal with / parse these errors?)
2. GET /projects/ end point, do we need it? Seems dangerous
3. Logging of project keys: We should never log these, we should treat them as a secret.
4. The code in Multi Mapeo essentially uses the project key as an identifier (for the instances and connections Maps). I think it would be better practice to use the discovery key as the "Project ID", and use this for in-memory Maps. The fewer places the project key is in the code, the fewer places we can accidentally leak it to logs or in responses.

File: index.js
Comment (RangerMauve) [reply index.js:46]: Raw sockets don't have an error type. The most we can do is end the socket connection.

I don't think there's an easy way to detect this error in the protocol, it would lead to an error being emitted in parsing the protocol data on the client side.

@noffle Do you know if the mapeo protocol has a place to handle error messages in the handshake? We could simulate one of those if they exist.

**Comment by RangerMauve** [comment]
Thanks for the review Gregor!

Here's some comments:

1. Fastify already responds with JSON when you reply with an error code. The docs for that are here: https://www.fastify.io/docs/latest/Reply/#errors
2. The /projects/ path is behind the same authentication as adding and removing projects. It was mostly there so that we could verify that stuff got added and to get a list of stuff that can be removed (as admins). I'm down to remove it, but I assumed that admins could be trusted with this route.
3. Should I log the discovery keys instead? Or maybe the first or last bytes of the project key the way mapeo-desktop and mapeo-mobile display them in the app?
4. I initially set that up that way because we need the project key when initializing a mapeo-core instance. This is also why replicate needs the project key. I see the benefit in using the discovery key in the map and within all the other methods though. Would leaving projectKey for get and replicate and using discovery key everywhere else work?

**Comment by gmaclennan** [comment]
> Thanks for the review Gregor!
> 
> Here's some comments:
> 
> 1. Fastify already responds with JSON when you reply with an error code. The docs for that are here: https://www.fastify.io/docs/latest/Reply/#errors

Oh that's great, that simplifies things a lot!

> 2. The /projects/ path is behind the same authentication as adding and removing projects. It was mostly there so that we could verify that stuff got added and to get a list of stuff that can be removed (as admins). I'm down to remove it, but I assumed that admins could be trusted with this route.

Makes sense. I think this should be a list of project ids, and I think we should use the discovery key as the project id. I know only admins can see this, but it means that we have a single point of failure (the basic auth password), that if someone gets access, then they can get access to all Mapeo projects around the world (because they would have all the encryption keys). I don't think even admins should be given an easy way to get everyones encryption keys.

> 3. Should I log the discovery keys instead? Or maybe the first or last bytes of the project key the way mapeo-desktop and mapeo-mobile display them in the app?

Let's use discovery keys for everything.

> 4. I initially set that up that way because we need the project key when initializing a mapeo-core instance. This is also why replicate needs the project key. I see the benefit in using the discovery key in the map and within all the other methods though. Would leaving projectKey for get and replicate and using discovery key everywhere else work?

Yes.

File: index.js
Comment (RangerMauve) [reply index.js]: Documented how? This could be any sort of uncaught exception. 🤔 Should the JSON structure for errors be documented in the README?

File: index.js
Comment (RangerMauve) [reply index.js]: I think I initially had an `await` in there, but that's since been removed. 😁 Good catch.

File: index.js
Comment (gmaclennan) [reply index.js:46]: I'd like to be clear about error handling before we merge this. What will the client do if it tries to sync and it is not "enabled"?
This is also important behaviour to test. Currently we have a test for successful sync. We should have a test for trying to sync when the project is not enabled on the server, to check that it fails correctly.

File: index.js
Comment (gmaclennan) [inline index.js:67]: Need to check that the project key is valid hex (`/[^0-9a-f]/.test(projectKey)`) because otherwise the server will crash with an uncaught error (because we do not catch the error that can be thrown in the discoveryKey function).

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Hi Mauve, thanks for these changes, all looking good. I had neglected to review the testing strategy last time, and I think it's important we test some key functionality before deploying this:

1. `test('Adding invalid project key (64 non-hex characters) fails gracefully')`
2. `test('Trying to sync with project key without permission fails')`
3. `test('Mapeo instance closes after timeout from last sync completion')`
4. `test('Mapeo instance does not close if a new sync starts after first completes but before GC timeout')`

File: index.js
Comment (RangerMauve) [reply index.js:67]: The error already throws properly, so I'm not sure if anything needs to be added here.

**Comment by RangerMauve** [comment]
I've added the extra unit tests you mentioned, Gregor. It seems that the behavior right now is that the connection closes before any sort of handshake can be done, so I'm not sure what else we could add to it.

Does this look good?

**Review by gmaclennan — APPROVED** [review]
Ok, I think this is good to go now.

**Comment by gmaclennan** [comment]
@mauve currently when I run `./bin.js list https://XXX:XXX@cloud.mapeo.app/` against the server I get a list with two items,  which is a duplicated project key. I would expect each project key to only exist once in the list.

**Comment by mauve** [comment]
I did get really confused by this notification :)

**Pull request closed without merging**
